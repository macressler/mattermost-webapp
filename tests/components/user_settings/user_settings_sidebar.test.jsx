// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React from 'react';
import {shallow} from 'enzyme';

import {mountWithIntl} from 'tests/helpers/intl-test-helper.jsx';

import UserSettingsSidebar from 'components/user_settings/sidebar/user_settings_sidebar.jsx';

describe('components/user_settings/sidebar/UserSettingsSidebar', () => {
    const defaultProps = {
        closeUnusedDirectMessages: 'after_seven_days',
        displayUnreadSection: 'true',
        showUnusedOption: false,
        showGroupSortOptions: true,
        sidebarPreference: {
            grouping: 'by_type',
            sorting: 'alpha',
            unreadsAtTop: 'true',
            favoriteAtTop: 'true',
        },
        user: {
            id: 'someuserid',
        },
        closeModal: () => () => true,
        collapseModal: () => () => true,
        updateSection: () => () => true,
        actions: {
            savePreferences: () => true,
        },
    };

    test('should match snapshot', () => {
        const wrapper = shallow(<UserSettingsSidebar {...defaultProps}/>);

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.state('isSaving')).toEqual(false);
        expect(wrapper.state('settings')).toEqual({
            close_unused_direct_messages: defaultProps.closeUnusedDirectMessages,
            grouping: defaultProps.sidebarPreference.grouping,
            sorting: defaultProps.sidebarPreference.sorting,
            unreads_at_top: defaultProps.sidebarPreference.unreadsAtTop,
            favorite_at_top: defaultProps.sidebarPreference.favoriteAtTop,
        });
    });

    test('should match state when updateSection is called', () => {
        const newUpdateSection = jest.fn();
        const updateArg = 'unreadChannels';
        const props = {...defaultProps, updateSection: newUpdateSection};
        const wrapper = shallow(<UserSettingsSidebar {...props}/>);

        wrapper.setState({isSaving: true});
        wrapper.instance().updateSection(updateArg);

        expect(wrapper.state('isSaving')).toEqual(false);
        expect(newUpdateSection).toHaveBeenCalledTimes(1);
        expect(newUpdateSection).toHaveBeenCalledWith(updateArg);
    });

    test('should pass handleChange for channel grouping', () => {
        const props = {...defaultProps, activeSection: 'groupChannels'};
        const wrapper = mountWithIntl(<UserSettingsSidebar {...props}/>);

        wrapper.find('#neverOption').simulate('change');
        expect(wrapper.state('settings')).toEqual({
            close_unused_direct_messages: defaultProps.closeUnusedDirectMessages,
            grouping: 'none',
            sorting: defaultProps.sidebarPreference.sorting,
            unreads_at_top: defaultProps.sidebarPreference.unreadsAtTop,
            favorite_at_top: defaultProps.sidebarPreference.favoriteAtTop,
        });

        wrapper.find('#byTypeOption').simulate('change');
        expect(wrapper.state('settings')).toEqual({
            close_unused_direct_messages: defaultProps.closeUnusedDirectMessages,
            grouping: 'by_type',
            sorting: defaultProps.sidebarPreference.sorting,
            unreads_at_top: defaultProps.sidebarPreference.unreadsAtTop,
            favorite_at_top: defaultProps.sidebarPreference.favoriteAtTop,
        });

        wrapper.find('#unreadAtTopOption').simulate('change', {target: {checked: false}});
        expect(wrapper.state('settings')).toEqual({
            close_unused_direct_messages: defaultProps.closeUnusedDirectMessages,
            grouping: 'by_type',
            sorting: defaultProps.sidebarPreference.sorting,
            unreads_at_top: 'false',
            favorite_at_top: defaultProps.sidebarPreference.favoriteAtTop,
        });

        wrapper.find('#unreadAtTopOption').simulate('change', {target: {checked: true}});
        expect(wrapper.state('settings')).toEqual({
            close_unused_direct_messages: defaultProps.closeUnusedDirectMessages,
            grouping: 'by_type',
            sorting: defaultProps.sidebarPreference.sorting,
            unreads_at_top: 'true',
            favorite_at_top: defaultProps.sidebarPreference.favoriteAtTop,
        });

        wrapper.find('#favoriteAtTopOption').simulate('change', {target: {checked: false}});
        expect(wrapper.state('settings')).toEqual({
            close_unused_direct_messages: defaultProps.closeUnusedDirectMessages,
            grouping: 'by_type',
            sorting: defaultProps.sidebarPreference.sorting,
            unreads_at_top: defaultProps.sidebarPreference.unreadsAtTop,
            favorite_at_top: 'false',
        });

        wrapper.find('#favoriteAtTopOption').simulate('change', {target: {checked: true}});
        expect(wrapper.state('settings')).toEqual({
            close_unused_direct_messages: defaultProps.closeUnusedDirectMessages,
            grouping: 'by_type',
            sorting: defaultProps.sidebarPreference.sorting,
            unreads_at_top: defaultProps.sidebarPreference.unreadsAtTop,
            favorite_at_top: 'true',
        });
    });

    test('should pass handleChange for channel grouping', () => {
        const props = {
            ...defaultProps,
            activeSection: 'sortChannels',
            sidebarPreference: {
                ...defaultProps.sidebarPreference,
                grouping: 'none',
            }};
        const wrapper = mountWithIntl(<UserSettingsSidebar {...props}/>);

        wrapper.find('#recentSectionEnabled').simulate('change');
        expect(wrapper.state('settings')).toEqual({
            close_unused_direct_messages: defaultProps.closeUnusedDirectMessages,
            grouping: 'none',
            sorting: 'recent',
            unreads_at_top: defaultProps.sidebarPreference.unreadsAtTop,
            favorite_at_top: defaultProps.sidebarPreference.favoriteAtTop,
        });

        wrapper.find('#alphaSectionEnabled').simulate('change');
        expect(wrapper.state('settings')).toEqual({
            close_unused_direct_messages: defaultProps.closeUnusedDirectMessages,
            grouping: 'none',
            sorting: 'alpha',
            unreads_at_top: defaultProps.sidebarPreference.unreadsAtTop,
            favorite_at_top: defaultProps.sidebarPreference.favoriteAtTop,
        });
    });
});
