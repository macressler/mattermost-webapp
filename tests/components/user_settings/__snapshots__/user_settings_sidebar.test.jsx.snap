// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/user_settings/sidebar/UserSettingsSidebar should match snapshot 1`] = `
<div>
  <div
    className="modal-header"
  >
    <button
      aria-label="Close"
      className="close"
      data-dismiss="modal"
      id="closeButton"
      onClick={[Function]}
      type="button"
    >
      <span
        aria-hidden="true"
      >
        Ã—
      </span>
    </button>
    <h4
      className="modal-title"
    >
      <div
        className="modal-back"
      >
        <i
          className="fa fa-angle-left"
          onClick={[Function]}
          title="Collapse Icon"
        />
      </div>
      <FormattedMessage
        defaultMessage="Sidebar Settings"
        id="user.settings.sidebar.title"
        values={Object {}}
      />
    </h4>
  </div>
  <div
    className="user-settings"
  >
    <h3
      className="tab-header"
    >
      <FormattedMessage
        defaultMessage="Sidebar Settings"
        id="user.settings.sidebar.title"
        values={Object {}}
      />
    </h3>
    <div
      className="divider-dark first"
    />
    <SettingItemMin
      describe={
        Array [
          <FormattedMessage
            defaultMessage="By Channel Type"
            id="user.settings.sidebar.channel_type"
            values={Object {}}
          />,
          <span>
            , 
          </span>,
          <FormattedMessage
            defaultMessage="Unreads and Favorites at the top"
            id="user.settings.sidebar.unreadsFavorites"
            values={Object {}}
          />,
        ]
      }
      focused={false}
      section="groupChannels"
      title={
        <FormattedMessage
          defaultMessage="Channel Grouping"
          id="user.settings.sidebar.groupChannelsTitle"
          values={Object {}}
        />
      }
      updateSection={[Function]}
    />
    <div
      className="divider-dark"
    />
    <SettingItemMin
      describe={
        <FormattedMessage
          defaultMessage="By Alphabetical Order"
          id="user.settings.sidebar.byAlpha"
          values={Object {}}
        />
      }
      focused={false}
      section="sortChannels"
      title={
        <FormattedMessage
          defaultMessage="Channel Sorting"
          id="user.settings.sidebar.sortChannelsTitle"
          values={Object {}}
        />
      }
      updateSection={[Function]}
    />
    <div
      className="divider-light"
    />
    <div
      className="divider-dark"
    />
  </div>
</div>
`;
